File: multimodal-generation/video-generation/javascript/amazonVideoUtil.js
import { promises as fs } from 'fs';
import path from 'path';
import { BedrockRuntimeClient, GetAsyncInvokeCommand, ListAsyncInvokesCommand } from '@aws-sdk/client-bedrock-runtime';
import { S3Client, ListObjectsV2Command, GetObjectCommand } from '@aws-sdk/client-s3';
import winston from 'winston';
import Ffmpeg from 'fluent-ffmpeg';

// Configure logger
const logger = winston.createLogger({
    level: 'info',
    format: winston.format.combine(
        winston.format.timestamp(),
        winston.format.json()
    ),
    transports: [
        new winston.transports.Console()
    ]
});

// Initialize AWS clients
const bedrockRuntime = new BedrockRuntimeClient();
const s3Client = new S3Client();

/**
 * Saves invocation information and model input to the specified output folder
 * @param {Object} invocationResult - The result from the async invocation
 * @param {Object} modelInput - The input provided to the model
 * @param {string} [outputFolder='output'] - The base output folder path
 * @returns {Promise<string>} The absolute path to the created output folder
 */
async function saveInvocationInfo(invocationResult, modelInput, outputFolder = 'output') {
    const invocationArn = invocationResult.invocationArn;
    logger.info(`Getting async invoke details for ARN: ${invocationArn}`);

    const command = new GetAsyncInvokeCommand({ invocationArn });
    const invocationJob = await bedrockRuntime.send(command);
    const folderName = getFolderNameForJob(invocationJob);

    const outputFolderAbs = path.resolve(`${outputFolder}/${folderName}`);
    await fs.mkdir(outputFolderAbs, { recursive: true });
    logger.info(`Saving invocation info to folder: ${outputFolderAbs}`);

    await fs.writeFile(
        path.join(outputFolderAbs, 'start_async_invoke_response.json'),
        JSON.stringify(invocationResult, null, 2)
    );
    logger.info('Saved start_async_invoke_response.json');

    await fs.writeFile(
        path.join(outputFolderAbs, 'model_input.json'),
        JSON.stringify(modelInput, null, 2)
    );
    logger.info('Saved model_input.json');

    return outputFolderAbs;
}

/**
 * Generates a folder name based on the invocation job details
 * @param {Object} invocationJob - The invocation job object
 * @param {string} invocationJob.invocationArn - The ARN of the invocation
 * @param {string} invocationJob.submitTime - The submission timestamp
 * @returns {string} The generated folder name
 */
function getFolderNameForJob(invocationJob) {
    const invocationArn = invocationJob.invocationArn;
    const invocationId = invocationArn.split('/').pop();
    const submitTime = invocationJob.submitTime;
    const timestamp = new Date(submitTime).toLocaleString('sv').replace(/[\s:]/g, '-');
    const folderName = `${timestamp}_${invocationId}`;
    logger.info(`Generated folder name: ${folderName}`);
    return folderName;
}

/**
 * Checks if a video file already exists for a given invocation job
 * @param {Object} invocationJob - The invocation job object
 * @param {string} [outputFolder='output'] - The base output folder path
 * @returns {Promise<boolean>} True if the video exists, false otherwise
 */
async function isVideoDownloadedForInvocationJob(invocationJob, outputFolder = 'output') {
    const invocationArn = invocationJob.invocationArn;
    const invocationId = invocationArn.split('/').pop();
    const folderName = getFolderNameForJob(invocationJob);
    const outputFolderAbs = path.resolve(`${outputFolder}/${folderName}`);
    const fileName = `${invocationId}.mp4`;
    const videoPath = path.join(outputFolderAbs, fileName);

    try {
        await fs.access(videoPath);
        return true;
    } catch (error) {
        return false;
    }
}

/**
 * Downloads a video file from S3 for a given invocation ARN
 * @param {string} invocationArn - The ARN of the invocation
 * @param {string} bucketName - The S3 bucket name
 * @param {string} destinationFolder - The folder where the video will be saved
 * @returns {Promise<string>} The local path to the downloaded video file
 */
async function downloadVideoForInvocationArn(invocationArn, bucketName, destinationFolder) {
    const invocationId = invocationArn.split('/').pop();
    const fileName = `${invocationId}.mp4`;
    const outputFolderAbs = path.resolve(destinationFolder);
    await fs.mkdir(outputFolderAbs, { recursive: true });
    const localFilePath = path.join(outputFolderAbs, fileName);
    logger.info(`Downloading video file to: ${localFilePath}`);

    const command = new ListObjectsV2Command({
        Bucket: bucketName,
        Prefix: invocationId
    });

    const response = await s3Client.send(command);

    for (const obj of response.Contents || []) {
        const objectKey = obj.Key;
        if (objectKey.endsWith('.mp4')) {
            const getObjectCommand = new GetObjectCommand({
                Bucket: bucketName,
                Key: objectKey
            });
            const { Body } = await s3Client.send(getObjectCommand);

            // Convert the readable stream to a buffer
            const chunks = [];
            for await (const chunk of Body) {
                chunks.push(chunk);
            }
            const buffer = Buffer.concat(chunks);

            // Write the buffer to file
            await fs.writeFile(localFilePath, buffer);

            logger.info(`Downloaded video file: ${localFilePath}`);
            return localFilePath;
        }
    }

    logger.error(`Problem: No MP4 file was found in S3 at ${bucketName}/${invocationId}`);
}

/**
 * Calculates the elapsed time for an invocation job
 * @param {Object} invocationJob - The invocation job object
 * @param {string} invocationJob.submitTime - The job submission time
 * @param {string} [invocationJob.endTime] - The job end time (if completed)
 * @returns {number} The elapsed time in seconds
 */
function elapsedTimeForInvocationJob(invocationJob) {
    const invocationStartTime = new Date(invocationJob.submitTime).getTime() / 1000;
    let elapsedTime;

    if (invocationJob.endTime) {
        const invocationEndTime = new Date(invocationJob.endTime).getTime() / 1000;
        elapsedTime = Math.floor(invocationEndTime - invocationStartTime);
    } else {
        elapsedTime = Math.floor(Date.now() / 1000 - invocationStartTime);
    }

    return elapsedTime;
}

/**
 * Logs the progress of a job including elapsed time
 * @param {Object} job - The job object to log progress for
 */
function logJobProgress(job) {
    const jobId = getJobIdFromArn(job.invocationArn);
    const elapsedTime = elapsedTimeForInvocationJob(job);
    const minutes = Math.floor(elapsedTime / 60);
    const seconds = elapsedTime % 60;
    logger.info(
        `Job ${jobId} is still in progress, elapsed time: ${minutes} minutes ${seconds} seconds`
    );
}

/**
 * Monitors and downloads a video for a specific invocation ARN
 * @param {string} invocationArn - The ARN of the invocation
 * @param {string} [outputFolder='output'] - The output folder path
 * @returns {Promise<string>} The local path to the downloaded video file
 */
async function monitorAndDownloadVideo(invocationArn, outputFolder = 'output') {
    logger.info(`Monitoring and downloading video for ARN: ${invocationArn}`);
    const command = new GetAsyncInvokeCommand({ invocationArn });
    const invocationJob = await bedrockRuntime.send(command);
    const status = invocationJob.status;

    let localFilePath;
    if (status === 'Completed') {
        localFilePath = await saveCompletedJob(invocationJob, outputFolder);
    } else if (status === 'Failed') {
        await saveFailedJob(invocationJob, outputFolder);
    } else {
        localFilePath = await monitorAndDownloadInProgressVideo(invocationArn, outputFolder);
    }

    return localFilePath;
}

/**
 * Monitors and downloads an in-progress video
 * @param {string} invocationArn - The ARN of the invocation
 * @param {string} [outputFolder='output'] - The output folder path
 * @returns {Promise<string>} The local path to the downloaded video file
 */
async function monitorAndDownloadInProgressVideo(invocationArn, outputFolder = 'output') {
    logger.info(`Monitoring and downloading in-progress video for ARN: ${invocationArn}`);
    let jobUpdate = await bedrockRuntime.send(new GetAsyncInvokeCommand({ invocationArn }));
    let status = jobUpdate.status;

    while (status === 'InProgress') {
        logJobProgress(jobUpdate);
        await new Promise(resolve => setTimeout(resolve, 10000)); // Sleep for 10 seconds
        jobUpdate = await bedrockRuntime.send(new GetAsyncInvokeCommand({ invocationArn }));
        status = jobUpdate.status;
    }

    if (status === 'Completed') {
        const localFilePath = await saveCompletedJob(jobUpdate, outputFolder);
        return localFilePath;
    } else if (status === 'Failed') {
        await saveFailedJob(jobUpdate, outputFolder);
    } else {
        logger.error(`Unexpected status: ${status} for job ${invocationArn}`);
    }
}

/**
 * Monitors and downloads all videos based on job status
 * @param {string} [outputFolder='output'] - The output folder path
 * @param {string} [submitTimeAfter=null] - Optional timestamp to filter jobs
 */
async function monitorAndDownloadVideos(outputFolder = 'output', submitTimeAfter = null) {
    logger.info('Monitoring and downloading videos');

    // Handle failed jobs
    const failedJobsArgs = { statusEquals: 'Failed' };
    if (submitTimeAfter) {
        failedJobsArgs.submitTimeAfter = submitTimeAfter;
    }

    const failedJobsCommand = new ListAsyncInvokesCommand(failedJobsArgs);
    const failedJobs = await bedrockRuntime.send(failedJobsCommand);
    for (const job of failedJobs.asyncInvokeSummaries) {
        await saveFailedJob(job, outputFolder);
    }

    // Handle completed jobs
    const completedJobsArgs = { statusEquals: 'Completed' };
    if (submitTimeAfter) {
        completedJobsArgs.submitTimeAfter = submitTimeAfter;
    }

    const completedJobsCommand = new ListAsyncInvokesCommand(completedJobsArgs);
    const completedJobs = await bedrockRuntime.send(completedJobsCommand);
    for (const job of completedJobs.asyncInvokeSummaries) {
        await saveCompletedJob(job, outputFolder);
    }

    await monitorAndDownloadInProgressVideos(outputFolder);
}

/**
 * Monitors and downloads videos for in-progress jobs
 * @param {string} [outputFolder='output'] - The output folder path
 */
async function monitorAndDownloadInProgressVideos(outputFolder = 'output') {
    logger.info('Monitoring and downloading in-progress videos');
    const listCommand = new ListAsyncInvokesCommand({ statusEquals: 'InProgress' });
    const invocationList = await bedrockRuntime.send(listCommand);
    const inProgressJobs = invocationList.asyncInvokeSummaries;
    let pendingJobArns = inProgressJobs.map(job => job.invocationArn);

    while (pendingJobArns.length > 0) {
        const jobArnsToRemove = [];

        for (const jobArn of pendingJobArns) {
            const command = new GetAsyncInvokeCommand({ invocationArn: jobArn });
            const jobUpdate = await bedrockRuntime.send(command);
            const status = jobUpdate.status;

            if (status === 'Completed' || status === 'Failed') {
                if (status === 'Completed') {
                    await saveCompletedJob(jobUpdate, outputFolder);
                } else {
                    await saveFailedJob(jobUpdate, outputFolder);
                }
                jobArnsToRemove.push(jobArn);
            } else {
                logJobProgress(jobUpdate);
            }
        }

        pendingJobArns = pendingJobArns.filter(arn => !jobArnsToRemove.includes(arn));
        await new Promise(resolve => setTimeout(resolve, 10000)); // Sleep for 10 seconds
    }

    logger.info('Monitoring and download complete!');
}

/**
 * Extracts the job ID from an invocation ARN
 * @param {string} invocationArn - The ARN of the invocation
 * @returns {string} The extracted job ID
 */
function getJobIdFromArn(invocationArn) {
    return invocationArn.split('/').pop();
}

/**
 * Saves details for a completed job and downloads its video
 * @param {Object} job - The completed job object
 * @param {string} [outputFolder='output'] - The output folder path
 * @returns {Promise<string>} The local path to the downloaded video file
 */
async function saveCompletedJob(job, outputFolder = 'output') {
    const jobId = getJobIdFromArn(job.invocationArn);
    const outputFolderAbs = path.resolve(`${outputFolder}/${getFolderNameForJob(job)}`);
    await fs.mkdir(outputFolderAbs, { recursive: true });
    logger.info(`Saving completed job details for job ${jobId}`);

    const statusFile = path.join(outputFolderAbs, 'completed.json');

    const videoExists = await isVideoDownloadedForInvocationJob(job, outputFolder);
    if (videoExists) {
        logger.info(`Skipping completed job ${jobId}, video already downloaded.`);
        return;
    }

    const s3BucketName = job.outputDataConfig.s3OutputDataConfig.s3Uri
        .split('//')[1]
        .split('/')[0];

    const localFilePath = await downloadVideoForInvocationArn(
        job.invocationArn,
        s3BucketName,
        outputFolderAbs
    );

    await fs.writeFile(statusFile, JSON.stringify(job, null, 2));
    logger.info(`Saved completed job details to ${statusFile}`);

    return localFilePath;
}

/**
 * Saves details for a failed job
 * @param {Object} job - The failed job object
 * @param {string} [outputFolder='output'] - The output folder path
 */
async function saveFailedJob(job, outputFolder = 'output') {
    const outputFolderAbs = path.resolve(`${outputFolder}/${getFolderNameForJob(job)}`);
    const outputFile = path.join(outputFolderAbs, 'failed.json');

    const jobId = getJobIdFromArn(job.invocationArn);

    try {
        await fs.access(outputFile);
        logger.info(`Skipping failed job ${jobId}, output file already exists.`);
        return;
    } catch (error) {
        await fs.mkdir(outputFolderAbs, { recursive: true });
        await fs.writeFile(outputFile, JSON.stringify(job, null, 2));
        logger.info(`Saved failed job details to ${outputFile}`);
    }
}

/**
 * Converts an image file to base64 string
 * @param {string} inputImagePath - The path to the input image file
 * @returns {Promise<string>} The base64 encoded string of the image
 * @throws {Error} If the image conversion fails
 */
async function imageToBase64(inputImagePath) {
    try {
        // Read the file
        const inputImageBytes = await fs.readFile(inputImagePath);

        // Convert to base64
        const inputImageBase64 = inputImageBytes.toString('base64');

        return inputImageBase64;
    } catch (error) {
        console.error('Error converting image to base64:', error);
        throw error;
    }
}

/**
 * Extracts the last frame of a video file.
 * 
 * @param {string} videoPath - The local path to the video to extract the last frame from.
 * @param {string} outputPath - The local path to save the extracted frame to.
 * @returns {Promise<void>}
 */
function extractLastFrame(videoPath, outputPath) {
    return new Promise((resolve, reject) => {
        // First, get the duration of the video
        Ffmpeg.ffprobe(videoPath, (err, metadata) => {
            if (err) {
                logger.error("Error: Could not open video.", err);
                return reject(err);
            }

            const duration = metadata.format.duration;

            // Extract the last frame by seeking to the end of the video
            Ffmpeg(videoPath)
                .seekInput(duration - 0.05) // Seek to just before the end
                .frames(1) // Extract only one frame
                .output(outputPath)
                .on('end', () => {
                    logger.info(`Last frame saved as ${outputPath}`);
                    resolve();
                })
                .on('error', (err) => {
                    logger.error("Error: Could not extract the last frame.", err);
                    reject(err);
                })
                .run();
        });
    });
}

/**
 * Concatenates two video files into a single output video
 * @param {string} video1Path - Path to the first video file
 * @param {string} video2Path - Path to the second video file
 * @param {string} outputPath - Path where the concatenated video should be saved
 * @returns {Promise} Resolves when concatenation is complete
 */
async function stitchVideos(video1Path, video2Path, outputPath) {
    try {
        // Check if input files exist using fs.access
        await Promise.all([
            fs.access(video1Path),
            fs.access(video2Path)
        ]);

        // Return a promise that resolves when the merge is complete
        return new Promise((resolve, reject) => {
            Ffmpeg(video1Path)
                .input(video2Path)
                .mergeToFile(outputPath)
                .on('end', () => {
                    logger.info('Files have been merged successfully');
                    resolve();
                })
                .on('error', (err) => {
                    logger.error('An error happened: ' + err.message);
                    reject(err);
                });
        });
    } catch (error) {
        throw error;
    }
}



export {
    saveInvocationInfo,
    getFolderNameForJob,
    isVideoDownloadedForInvocationJob,
    downloadVideoForInvocationArn,
    elapsedTimeForInvocationJob,
    monitorAndDownloadVideo,
    monitorAndDownloadVideos,
    monitorAndDownloadInProgressVideos,
    saveCompletedJob,
    saveFailedJob,
    getJobIdFromArn,
    imageToBase64,
    extractLastFrame,
    stitchVideos
};

File: multimodal-generation/video-generation/python/amazon_video_util.py
import os
import time
import boto3
import json
import logging
import cv2
from datetime import datetime
from moviepy import VideoFileClip, CompositeVideoClip

logger = logging.getLogger(__name__)

bedrock_runtime = boto3.client("bedrock-runtime")
s3_client = boto3.client("s3")


def save_invocation_info(
    invocation_result: dict, model_input: dict, output_folder: str = "output"
) -> str:
    """
    Save the invocation result and model input to the specified output folder.

    Args:
        invocation_result (dict): The result of the video generation invocation.
        model_input (dict): The input parameters used for the video generation.
        output_folder (str, optional): The folder where the invocation info will be saved. Defaults to "output".

    Returns:
        str: The absolute path of the output folder.
    """
    invocation_arn = invocation_result["invocationArn"]
    logger.info(f"Getting async invoke details for ARN: {invocation_arn}")
    invocation_job = bedrock_runtime.get_async_invoke(invocationArn=invocation_arn)
    folder_name = get_folder_name_for_job(invocation_job)

    output_folder_abs = os.path.abspath(f"{output_folder}/{folder_name}")
    os.makedirs(output_folder_abs, exist_ok=True)
    logger.info(f"Saving invocation info to folder: {output_folder_abs}")

    with open(
        os.path.join(output_folder_abs, "start_async_invoke_response.json"), "w"
    ) as f:
        json.dump(invocation_result, f, indent=2, default=str)
        logger.info("Saved start_async_invoke_response.json")

    with open(os.path.join(output_folder_abs, "model_input.json"), "w") as f:
        json.dump(model_input, f, indent=2, default=str)
        logger.info("Saved model_input.json")

    return output_folder_abs


def get_folder_name_for_job(invocation_job: dict) -> str:
    """
    Generate a folder name for the job based on the invocation ARN and submission time.

    Args:
        invocation_job (dict): The job details containing the invocation ARN and submission time.

    Returns:
        str: The generated folder name.
    """
    invocation_arn = invocation_job["invocationArn"]
    invocation_id = invocation_arn.split("/")[-1]
    submit_time = invocation_job["submitTime"]
    timestamp = submit_time.astimezone().strftime("%Y-%m-%d_%H-%M-%S")
    folder_name = f"{timestamp}_{invocation_id}"
    logger.info(f"Generated folder name: {folder_name}")
    return folder_name


def is_video_downloaded_for_invocation_job(
    invocation_job: dict, output_folder: str = "output"
) -> bool:
    """
    Check if the video file for the given invocation job has been downloaded.

    Args:
        invocation_job (dict): The job details containing the invocation ARN.
        output_folder (str, optional): The folder where the video is expected to be downloaded. Defaults to "output".

    Returns:
        bool: True if the video file exists, False otherwise.
    """
    invocation_arn = invocation_job["invocationArn"]
    invocation_id = invocation_arn.split("/")[-1]
    folder_name = get_folder_name_for_job(invocation_job)
    output_folder_abs = os.path.abspath(f"{output_folder}/{folder_name}")
    file_name = f"{invocation_id}.mp4"
    video_path = os.path.join(output_folder_abs, file_name)
    if os.path.exists(video_path):
        logger.info(f"Video file already downloaded: {video_path}")
        return True
    logger.info(f"Video file not found: {video_path}")
    return False


def download_video_for_invocation_arn(
    invocation_arn: str, bucket_name: str, destination_folder: str
):
    """
    Download the video file for the given invocation ARN from the specified S3 bucket.

    Args:
        invocation_arn (str): The ARN of the invocation.
        bucket_name (str): The name of the S3 bucket.
        destination_folder (str): The local folder where the video will be downloaded.

    Returns:
        str: Local file path for the downloaded video.
    """
    invocation_id = invocation_arn.split("/")[-1]
    file_name = f"{invocation_id}.mp4"
    output_folder_abs = os.path.abspath(destination_folder)
    os.makedirs(output_folder_abs, exist_ok=True)
    local_file_path = os.path.join(output_folder_abs, file_name)
    logger.info(f"Downloading video file to: {local_file_path}")

    response = s3_client.list_objects_v2(Bucket=bucket_name, Prefix=invocation_id)

    for obj in response.get("Contents", []):
        object_key = obj["Key"]
        if object_key.endswith(".mp4"):
            s3_client.download_file(bucket_name, object_key, local_file_path)
            logger.info(f"Downloaded video file: {local_file_path}")
            return local_file_path

    logger.error(
        f"Problem: No MP4 file was found in S3 at {bucket_name}/{invocation_id}"
    )


def elapsed_time_for_invocation_job(invocation_job: dict) -> int:
    """
    Calculate the elapsed time for the given invocation job.

    Args:
        invocation_job (dict): The job details containing the submission and end times.

    Returns:
        int: The elapsed time in seconds.
    """
    invocation_start_time = invocation_job["submitTime"].timestamp()
    if "endTime" in invocation_job:
        invocation_end_time = invocation_job["endTime"].timestamp()
        elapsed_time = int(invocation_end_time - invocation_start_time)
    else:
        elapsed_time = int(time.time() - invocation_start_time)

    return elapsed_time


def log_job_progress(job: dict):
    """
    Helper function to log job progress.

    Args:
        job (dict): The job details containing the invocation ARN.
    """
    job_id = get_job_id_from_arn(job["invocationArn"])
    elapsed_time = elapsed_time_for_invocation_job(job)
    minutes, seconds = divmod(elapsed_time, 60)
    logger.info(
        f"Job {job_id} is still in progress, elapsed time: {minutes} minutes {seconds} seconds"
    )


def monitor_and_download_video(invocation_arn: str, output_folder: str = "output"):
    """
    Monitor and download a single video based on the provided invocation ARN.

    Args:
        invocation_arn (str): The ARN of the invocation.
        output_folder (str, optional): The folder where the video will be downloaded. Defaults to "output".

    Returns:
        str: Local file path for the downloaded video.
    """
    logger.info(f"Monitoring and downloading video for ARN: {invocation_arn}")
    invocation_job = bedrock_runtime.get_async_invoke(invocationArn=invocation_arn)
    status = invocation_job["status"]

    if status == "Completed":
        local_file_path = save_completed_job(
            invocation_job, output_folder=output_folder
        )
    elif status == "Failed":
        save_failed_job(invocation_job, output_folder=output_folder)
    else:
        local_file_path = monitor_and_download_in_progress_video(
            invocation_arn, output_folder=output_folder
        )

    return local_file_path


def monitor_and_download_in_progress_video(
    invocation_arn: str, output_folder: str = "output"
):
    """
    Monitor and download a single video that is currently in progress.

    Args:
        invocation_arn (str): The ARN of the invocation.
        output_folder (str, optional): The folder where the video will be downloaded. Defaults to "output".

    Returns:
        str: Local file path for the downloaded video.
    """
    logger.info(
        f"Monitoring and downloading in-progress video for ARN: {invocation_arn}"
    )
    job_update = bedrock_runtime.get_async_invoke(invocationArn=invocation_arn)
    status = job_update["status"]

    while status == "InProgress":
        log_job_progress(job_update)
        time.sleep(10)
        job_update = bedrock_runtime.get_async_invoke(invocationArn=invocation_arn)
        status = job_update["status"]

    if status == "Completed":
        local_file_path = save_completed_job(job_update, output_folder=output_folder)
        return local_file_path
    elif status == "Failed":
        save_failed_job(job_update, output_folder=output_folder)
    else:
        logger.error(f"Unexpected status: {status} for job {invocation_arn}")


def monitor_and_download_videos(
    output_folder: str = "output", submit_time_after: datetime = None
):
    """
    Monitor and download videos for jobs that are in progress, completed, or failed.

    Args:
        output_folder (str, optional): The folder where the videos will be downloaded. Defaults to "output".
        submit_time_after (datetime, optional): Filter jobs submitted after this time.
    """
    logger.info("Monitoring and downloading videos")
    failed_jobs_args = {"statusEquals": "Failed"}
    if submit_time_after:
        failed_jobs_args["submitTimeAfter"] = submit_time_after

    failed_jobs = bedrock_runtime.list_async_invokes(**failed_jobs_args)
    for job in failed_jobs["asyncInvokeSummaries"]:
        save_failed_job(job, output_folder=output_folder)

    completed_jobs_args = {"statusEquals": "Completed"}
    if submit_time_after:
        completed_jobs_args["submitTimeAfter"] = submit_time_after

    completed_jobs = bedrock_runtime.list_async_invokes(**completed_jobs_args)
    for job in completed_jobs["asyncInvokeSummaries"]:
        save_completed_job(job, output_folder=output_folder)

    monitor_and_download_in_progress_videos(output_folder=output_folder)


def monitor_and_download_in_progress_videos(output_folder: str = "output"):
    """
    Monitor and download videos for jobs that are currently in progress.

    Args:
        output_folder (str, optional): The folder where the videos will be downloaded. Defaults to "output".
    """
    logger.info("Monitoring and downloading in-progress videos")
    invocation_list = bedrock_runtime.list_async_invokes(statusEquals="InProgress")
    in_progress_jobs = invocation_list["asyncInvokeSummaries"]
    pending_job_arns = [job["invocationArn"] for job in in_progress_jobs]

    while pending_job_arns:
        job_arns_to_remove = []
        for job_arn in pending_job_arns:
            job_update = bedrock_runtime.get_async_invoke(invocationArn=job_arn)
            status = job_update["status"]

            if status in ("Completed", "Failed"):
                if status == "Completed":
                    save_completed_job(job_update, output_folder=output_folder)
                else:
                    save_failed_job(job_update, output_folder=output_folder)
                job_arns_to_remove.append(job_arn)
            else:
                log_job_progress(job_update)

        for job_arn in job_arns_to_remove:
            pending_job_arns.remove(job_arn)

        time.sleep(10)

    logger.info("Monitoring and download complete!")


def get_job_id_from_arn(invocation_arn: str) -> str:
    """
    Extract the job ID from the invocation ARN.

    Args:
        invocation_arn (str): The invocation ARN.

    Returns:
        str: The job ID.
    """
    return invocation_arn.split("/")[-1]


def save_completed_job(job: dict, output_folder: str = "output"):
    """
    Save the details of a completed job and download the video if it hasn't been downloaded yet.

    Args:
        job (dict): The job details.
        output_folder (str, optional): The folder where the job details and video will be saved. Defaults to "output".

    Returns:
        str: Local file path for the downloaded video.
    """
    job_id = get_job_id_from_arn(job["invocationArn"])
    output_folder_abs = os.path.abspath(
        f"{output_folder}/{get_folder_name_for_job(job)}"
    )
    os.makedirs(output_folder_abs, exist_ok=True)
    logger.info(f"Saving completed job details for job {job_id}")

    status_file = os.path.join(output_folder_abs, "completed.json")

    if is_video_downloaded_for_invocation_job(job, output_folder=output_folder):
        logger.info(f"Skipping completed job {job_id}, video already downloaded.")
        return

    s3_bucket_name = (
        job["outputDataConfig"]["s3OutputDataConfig"]["s3Uri"]
        .split("//")[1]
        .split("/")[0]
    )
    local_file_path = download_video_for_invocation_arn(
        job["invocationArn"], s3_bucket_name, output_folder_abs
    )

    with open(status_file, "w") as f:
        json.dump(job, f, indent=2, default=str)
        logger.info(f"Saved completed job details to {status_file}")

    return local_file_path


def save_failed_job(job: dict, output_folder: str = "output"):
    """
    Save the details of a failed job.

    Args:
        job (dict): The job details.
        output_folder (str, optional): The folder where the job details will be saved. Defaults to "output".
    """
    output_folder_abs = os.path.abspath(
        f"{output_folder}/{get_folder_name_for_job(job)}"
    )
    output_file = os.path.join(output_folder_abs, "failed.json")

    job_id = get_job_id_from_arn(job["invocationArn"])

    if os.path.exists(output_file):
        logger.info(f"Skipping failed job {job_id}, output file already exists.")
        return

    os.makedirs(output_folder_abs, exist_ok=True)

    with open(output_file, "w") as f:
        json.dump(job, f, indent=2, default=str)
        logger.info(f"Saved failed job details to {output_file}")


def extract_last_frame(video_path: str, output_path: str):
    """
    Extracts the last frame of a video file.

    Args:
        video_path (str): The local path to the video to extract the last frame from.
        output_path (str): The local path to save the extracted frame to.
    """
    # Open the video file
    cap = cv2.VideoCapture(video_path)

    # Check if the video file is opened successfully
    if not cap.isOpened():
        logger.error("Error: Could not open video.")
        return

    # Get the total number of frames in the video
    total_frames = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))

    # Move to the last frame
    cap.set(cv2.CAP_PROP_POS_FRAMES, total_frames - 1)

    # Read the last frame
    ret, frame = cap.read()

    # Check if the frame is read successfully
    if ret:
        # Save the last frame as an image
        cv2.imwrite(output_path, frame)
        logger.info(f"Last frame saved as {output_path}")
    else:
        logger.error("Error: Could not read the last frame.")

    # Release the video capture object
    cap.release()


def stitch_videos(video1_path: str, video2_path: str, output_path: str):
    """
    Stitches two videos together and saves the result to a new file.

    Args:
        video1_path (str): The file path to the first video.
        video2_path (str): The file path to the second video.
        output_path (str): The file path to save the stitched video.
    """
    # Load the video clips
    clip1 = VideoFileClip(video1_path)
    clip2 = VideoFileClip(video2_path)

    final_clip = [
        clip1,
        clip2.with_start(clip1.duration),
    ]

    # Concatenate the clips
    final_clip = CompositeVideoClip(final_clip)

    # Write the result
    final_clip.write_videofile(output_path)

    # Clean up
    clip1.close()
    clip2.close()
    final_clip.close()

    logger.info(f"Stitched video saved to {output_path}")
